src/Java.Interop.Tools.Maven/Models/ResolvedProject.cs

```diff
@@ -0,0 +1,186 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Java.Interop.Tools.Maven.Extensions;
+
+namespace Java.Interop.Tools.Maven.Models;
+
+public class ResolvedProject
+{
+	readonly ResolvedProject? parent;
+	readonly IPomResolver? resolver;
+
+	Project? resolved_project;
+
+	public Project Raw { get; }
+	public Project Resolved => resolved_project ?? throw new InvalidOperationException ("Call the Resolve method before accessing this.");
+
+	public virtual bool IsSuperPom => false;
+	public string ArtifactId => Resolved.ArtifactId.OrEmpty ();
+	public string GroupId => Resolved.GroupId.HasValue () || IsSuperPom ? Resolved.GroupId.OrEmpty () : Parent.GroupId;
+	public string Version => Resolved.Version.HasValue () || IsSuperPom ? Resolved.Version.OrEmpty () : Parent.Version;
+	public string Name => Resolved.Name.HasValue () || IsSuperPom ? Resolved.Name.OrEmpty () : Parent.Name;
+
+	public List<ResolvedDependency> Dependencies { get; } = new List<ResolvedDependency> ();
+	public List<ResolvedProject> ImportedPomProjects { get; } = new (); // Projects imported via a scope = "import", type = "pom" dependencyManagement section
+
+	public ResolvedProject Parent {
+		get {
+			if (parent is null && IsSuperPom)
+				throw new InvalidOperationException ("Super POM does not have a parent, check IsSuperPom before calling");
+
+			return parent ?? throw new InvalidOperationException ("Parent should not be null");
+		}
+	}
+
+	public ResolvedProject (Project project, ResolvedProject parent, IPomResolver resolver)
+	{
+		Raw = project;
+		this.parent = parent;
+		this.resolver = resolver;
+	}
+
+	public void Resolve () => ResolveCore (new PropertyStack ());
+
+	public static ResolvedProject FromArtifact (Artifact artifact, IPomResolver resolver)
+	{
+		var project = FromArtifactCore (artifact, resolver);
+		project.Resolve ();
+
+		return project;
+	}
+
+	static ResolvedProject FromArtifactCore (Artifact artifact, IPomResolver resolver)
+	{
+		var raw = resolver.ResolveRawProject (artifact);
+
+		// POM has a parent, resolve it
+		if (raw.TryGetParentPomArtifact (out var parentArtifact)) {
+			var parent = FromArtifactCore (parentArtifact, resolver);
+			return new ResolvedProject (raw, parent, resolver);
+		}
+
+		var project = new ResolvedProject (raw, SuperPom.Instance, resolver);
+
+		return project;
+	}
+
+	void ResolveCore (PropertyStack properties)
+	{
+		if (IsSuperPom)
```

Comment:
Why is it called a "super POM" instead of a "root POM"?  Maybe it's just me, but "super" doesn't imply (to me) that it has no parents; "root" does.  (`/` has no parent directory, `C:\` has no parent directory, etc.)

Reply:
…and "super POM" might be "better" in that it's the existing terminology of the space, but it doesn't feel "self documenting", as it were.

Reply:
Correct, using Maven's terminology for this concept:
https://maven.apache.org/guides/introduction/introduction-to-the-pom.html#super-pom